hive-metastore/src/main/java/org/apache/iceberg/hive/ClientPoolImpl.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.hive;
+
+import java.io.Closeable;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class ClientPoolImpl<C, E extends Exception> implements Closeable, ClientPool<C, E> {
```

Comment:
@lcspinter, @pvary, why was it necessary to create an abstract client pool implementation?

Reply:
@rdblue The `ClientPoolImpl` is basically just the renamed copy of `ClientPool` from the previous version of the code, with a slight modification. What changed is that the `Action` interface and `run` method was extracted into a separate `ClientPool` interface (That's why we had to do the renaming). This interface is implemented by the `CacheClientPool` and `ClientPoolImpl` classes (and indirectly by `HiveClientPool`). 
When someone calls `run` on an instance of `CachedClientPool`, we just fetch the appropriate `HiveClientPool` instance from the cache and call `run` on that instance.

Reply:
Thanks, @lcspinter! The context I was missing was that the `CacheClientPool` implements the interface.